File: src/JsonServices.Core/Events/ClientSubscription.cs
using System;
using System.Collections.Generic;

namespace JsonServices.Events
{
	internal class ClientSubscription<TEventArgs> : IClientSubscription
		where TEventArgs : EventArgs
	{
		public string SubscriptionId { get; set; }

		public string EventName { get; set; }

		public EventHandler<TEventArgs> EventHandler { get; set; }

		public Dictionary<string, string> EventFilter { get; set; }

		public void Invoke(object sender, EventArgs args)
		{
			if (EventHandler != null && EventFilter.Matches(args))
			{
				EventHandler?.Invoke(sender, (TEventArgs)args);
			}
		}

		private SubscriptionMessage.Subscription CreateSubscription()
		{
			return new SubscriptionMessage.Subscription
			{
				SubscriptionId = SubscriptionId,
				EventName = EventName,
				EventFilter = EventFilter,
				Enabled = true,
			};
		}

		public SubscriptionMessage CreateSubscriptionMessage()
		{
			return new SubscriptionMessage
			{
				Subscriptions = new[] { CreateSubscription() },
			};
		}

		public SubscriptionMessage CreateUnsubscriptionMessage()
		{
			var sub = CreateSubscription();
			sub.EventFilter = null; // no need to send the filter again
			sub.Enabled = false;

			return new SubscriptionMessage
			{
				Subscriptions = new[] { sub },
			};
		}
	}
}

File: sample/sample-client/src/JsonServices/ClientSubscription.ts
import { EventFilter } from './EventFilter';
import { SubscriptionMessage } from './SubscriptionMessage';

export class ClientSubscription {
    public subscriptionId: string;
    public eventName: string;
    public eventHandler: (eventArgs: object) => void;
    public eventFilter?: {
        [key: string]: string;
    }

    public invoke = (eventArgs: object) => {
        // TODO: handle 'this' context?
        // apply eventFilter locally (we might get events matching other local subscriber's event filter)
        if (EventFilter.matches(this.eventFilter, eventArgs)) {
            this.eventHandler(eventArgs);
        }
    }

    public createSubscriptionMessage = () => {
        const msg = new SubscriptionMessage();
        msg.Subscriptions = [{
            Enabled: true,
            EventName: this.eventName,
            EventFilter: this.eventFilter,
            SubscriptionId: this.subscriptionId,
        }];
        return msg;
    }

    public createUnsubscriptionMessage = () => {
        const msg = this.createSubscriptionMessage();
        delete msg.Subscriptions[0].EventFilter;
        msg.Subscriptions[0].Enabled = false;
        return msg;
    }
}